File: Data Structures & Algorithms/decode-ways/submission-0.java
class Solution {

    public int numDecodings(String s) {
        return dfs(s, 0);
    }

    private int dfs(String s, int i){
        if (s.length() == i ) return 1;
        if ( s.charAt(i) == '0') return 0;

        int res = dfs(s, i+1);

        if (i < s.length() - 1){
            if (s.charAt(i) == '1' || (s.charAt(i) <= '2' && s.charAt(i+1) < '7') ){
                res += dfs(s, i+2);
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) {
        candidates.sort((a, b) => a - b);
        const res  = [];
        this.dfs(candidates, target, 0, [], 0, res);
        return res;
    }

    dfs (candidates, target, i , cur, total, res){
        if (target === total){
            res.push([...cur]);
            return;
        }

        if (total > target || i === candidates.length) return;

        cur.push(candidates[i]);
        this.dfs(candidates, target, i+1, cur, total + candidates[i], res);
        cur.pop();

        while (i+1 < candidates.length && candidates[i] === candidates[i+1]){
            i++;
        }

        this.dfs(candidates , target, i+1, cur, total, res);
    }
}
